File: backend/app/schemas/developer.py
"""Pydantic schemas for the Developer API surface."""

from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field, field_validator

from app.models.developer import DeveloperScope, EventSeverity

# ──────────────────────────────────────────────────────────────────────
# Developer API Keys (extends existing api_keys with scopes)
# ──────────────────────────────────────────────────────────────────────


class DeveloperAPIKeyCreate(BaseModel):
    """Body for POST /developer/keys."""

    name: str = Field(..., min_length=1, max_length=100)
    description: str | None = None
    scopes: list[DeveloperScope] = Field(
        default_factory=lambda: [DeveloperScope.EVENTS_WRITE],
        description="Permissions to grant. Defaults to events:write.",
    )

    @field_validator("scopes")
    @classmethod
    def at_least_one_scope(cls, v: list[DeveloperScope]) -> list[DeveloperScope]:
        if not v:
            raise ValueError("At least one scope is required")
        return v


class DeveloperAPIKeyResponse(BaseModel):
    """A key as returned to the UI (full plaintext key is returned only once
    on create — see DeveloperAPIKeyCreateResponse)."""

    id: str
    name: str
    description: str | None
    key_prefix: str
    is_active: bool
    scopes: list[DeveloperScope]
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class DeveloperAPIKeyCreateResponse(DeveloperAPIKeyResponse):
    """Returned by POST /developer/keys — includes the plaintext key, shown once."""

    full_key: str = Field(..., description="Plaintext key — store immediately, shown only once.")


class DeveloperAPIKeyListResponse(BaseModel):
    keys: list[DeveloperAPIKeyResponse]
    total: int


class DeveloperAPIKeyScopesUpdate(BaseModel):
    """Body for PATCH /developer/keys/{id}/scopes."""

    scopes: list[DeveloperScope]

    @field_validator("scopes")
    @classmethod
    def at_least_one_scope(cls, v: list[DeveloperScope]) -> list[DeveloperScope]:
        if not v:
            raise ValueError("At least one scope is required")
        return v


# ──────────────────────────────────────────────────────────────────────
# Event Definitions
# ──────────────────────────────────────────────────────────────────────


class EventDefinitionCreate(BaseModel):
    # `protected_namespaces=()` silences Pydantic's warning about `schema`
    # shadowing BaseModel.schema; we genuinely want a `schema` field on the
    # wire because that's what callers will write in their API requests.
    model_config = {"protected_namespaces": ()}

    name: str = Field(..., min_length=1, max_length=120, pattern=r"^[a-z][a-z0-9._-]*$")
    description: str | None = None
    schema: dict[str, Any] | None = Field(
        default=None,
        description="Advisory schema — keys are field names, values are types ('string', 'integer', etc.).",
    )
    default_severity: EventSeverity = EventSeverity.INFO


class EventDefinitionUpdate(BaseModel):
    model_config = {"protected_namespaces": ()}

    description: str | None = None
    schema: dict[str, Any] | None = None
    is_enabled: bool | None = None
    default_severity: EventSeverity | None = None


class EventDefinitionResponse(BaseModel):
    model_config = {"protected_namespaces": (), "from_attributes": True}

    id: str
    name: str
    description: str | None
    schema: dict[str, Any] | None
    is_enabled: bool
    default_severity: EventSeverity
    created_at: datetime
    updated_at: datetime


class EventDefinitionListResponse(BaseModel):
    definitions: list[EventDefinitionResponse]
    total: int


# ──────────────────────────────────────────────────────────────────────
# Event Ingestion + Listing
# ──────────────────────────────────────────────────────────────────────


class EventIngestItem(BaseModel):
    """A single event in an ingestion batch."""

    event_type: str = Field(..., min_length=1, max_length=120)
    severity: EventSeverity | None = None
    occurred_at: datetime | None = None
    payload: dict[str, Any] = Field(default_factory=dict)
    source: str | None = Field(default=None, max_length=120)
    session_id: str | None = Field(default=None, max_length=120)
    user_external_id: str | None = Field(default=None, max_length=255)
    correlation_id: str | None = Field(default=None, max_length=120)
    occurrences: int = Field(default=1, ge=1, le=10_000)


class EventIngestRequest(BaseModel):
    """POST /developer/events — batched ingestion."""

    events: list[EventIngestItem] = Field(..., min_length=1, max_length=500)


class EventIngestResponse(BaseModel):
    ingested: int
    skipped: int = 0
    skipped_reasons: list[str] = Field(default_factory=list)


class EventResponse(BaseModel):
    id: str
    event_type: str
    severity: EventSeverity
    occurred_at: datetime
    payload: dict[str, Any]
    source: str | None
    session_id: str | None
    user_external_id: str | None
    correlation_id: str | None
    occurrences: int
    created_at: datetime

    class Config:
        from_attributes = True


class EventListResponse(BaseModel):
    events: list[EventResponse]
    total: int
    has_more: bool
    next_cursor: str | None = None

File: frontend/src/lib/developer.ts
'use client';

/**
 * Developer API client and types.
 *
 * Mirrors the backend Pydantic schemas in app/schemas/developer.py.
 *
 * Authentication: delegates to the shared ApiClient in `api.ts`, which
 * handles the Authorization header and 401 refresh-and-retry.
 */

import { api } from './api';

// ──────────────────────────────────────────────────────────────────────
// Types
// ──────────────────────────────────────────────────────────────────────

export type DeveloperScope =
  | 'events:write'
  | 'events:read'
  | 'event_types:write'
  | 'event_types:read';

export type EventSeverity =
  | 'debug'
  | 'info'
  | 'warning'
  | 'error'
  | 'critical';

export const ALL_SCOPES: DeveloperScope[] = [
  'events:write',
  'events:read',
  'event_types:write',
  'event_types:read',
];

export const ALL_SEVERITIES: EventSeverity[] = [
  'debug',
  'info',
  'warning',
  'error',
  'critical',
];

export interface DeveloperAPIKey {
  id: string;
  name: string;
  description: string | null;
  key_prefix: string;
  is_active: boolean;
  scopes: DeveloperScope[];
  created_at: string;
  updated_at: string;
}

export interface DeveloperAPIKeyCreateResponse extends DeveloperAPIKey {
  full_key: string;
}

export interface EventDefinition {
  id: string;
  name: string;
  description: string | null;
  schema: Record<string, unknown> | null;
  is_enabled: boolean;
  default_severity: EventSeverity;
  created_at: string;
  updated_at: string;
}

export interface DeveloperEvent {
  id: string;
  event_type: string;
  severity: EventSeverity;
  occurred_at: string;
  payload: Record<string, unknown>;
  source: string | null;
  session_id: string | null;
  user_external_id: string | null;
  correlation_id: string | null;
  occurrences: number;
  created_at: string;
}

export interface EventListResponse {
  events: DeveloperEvent[];
  total: number;
  has_more: boolean;
  next_cursor: string | null;
}

// ──────────────────────────────────────────────────────────────────────
// Developer API surface
// ──────────────────────────────────────────────────────────────────────

export const developerApi = {
  // ── API Keys ─────────────────────────────────────────────────────
  listKeys: (includeInactive = false) =>
    api.request<{ keys: DeveloperAPIKey[]; total: number }>(
      `/developer/keys?include_inactive=${includeInactive}`,
    ),

  createKey: (body: {
    name: string;
    description?: string | null;
    scopes?: DeveloperScope[];
  }) =>
    api.request<DeveloperAPIKeyCreateResponse>('/developer/keys', {
      method: 'POST',
      body: JSON.stringify(body),
    }),

  revokeKey: (keyId: string) =>
    api.request<void>(`/developer/keys/${keyId}`, { method: 'DELETE' }),

  updateKeyScopes: (keyId: string, scopes: DeveloperScope[]) =>
    api.request<DeveloperAPIKey>(`/developer/keys/${keyId}/scopes`, {
      method: 'PATCH',
      body: JSON.stringify({ scopes }),
    }),

  // ── Event Types ──────────────────────────────────────────────────
  listEventTypes: (includeDisabled = false) =>
    api.request<{ definitions: EventDefinition[]; total: number }>(
      `/developer/event-types?include_disabled=${includeDisabled}`,
    ),

  createEventType: (body: {
    name: string;
    description?: string | null;
    schema?: Record<string, unknown> | null;
    default_severity?: EventSeverity;
  }) =>
    api.request<EventDefinition>('/developer/event-types', {
      method: 'POST',
      body: JSON.stringify(body),
    }),

  updateEventType: (
    id: string,
    body: Partial<{
      description: string | null;
      schema: Record<string, unknown> | null;
      is_enabled: boolean;
      default_severity: EventSeverity;
    }>,
  ) =>
    api.request<EventDefinition>(`/developer/event-types/${id}`, {
      method: 'PATCH',
      body: JSON.stringify(body),
    }),

  deleteEventType: (id: string) =>
    api.request<void>(`/developer/event-types/${id}`, { method: 'DELETE' }),

  // ── Events ───────────────────────────────────────────────────────
  listEvents: (params?: {
    event_type?: string;
    severity?: EventSeverity;
    source?: string;
    session_id?: string;
    correlation_id?: string;
    since?: string;
    until?: string;
    limit?: number;
    after?: string;
  }) => {
    const q = new URLSearchParams();
    if (params) {
      for (const [key, value] of Object.entries(params)) {
        if (value !== undefined && value !== null && value !== '') {
          q.set(key, String(value));
        }
      }
    }
    return api.request<EventListResponse>(
      `/developer/events${q.toString() ? `?${q}` : ''}`,
    );
  },

  getEvent: (id: string) =>
    api.request<DeveloperEvent>(`/developer/events/${id}`),
};
